### PR TITLE
allow datasources to have onAfterSave hook

### PR DESCRIPTION
**Pull request by svenklemm**
This PR adds an onAfterSave hook which will trigger once a datasource has been saved and before testDatasource.
The hook gets the current configuration of the datasource as argument and may change settings. By returning {changed: true} datasource edit control will save the datasource again to persist settings.

#12680

### CODE REVIEW

File: public/app/features/plugins/ds_edit_ctrl.ts [hunk 1]
```diff
@@ -120,6 +122,27 @@ export class DataSourceEditCtrl {
...
+    return this.datasourceSrv.get(this.current.name).then(datasource => {
+      if (!datasource.onAfterSave) {
+        return;
+      }
+
+      return datasource
+        .onAfterSave(this.current)
+        .then(result => {
+          if (result.changed === true) {
+            return this.backendSrv.put('/api/datasources/' + this.current.id, this.current).then(result => {
```
Comment (solugebefola) [inline public/app/features/plugins/ds_edit_ctrl.ts]: I would just say to place the `.then` statement on the next line to be consistent.

**Comment by marefr** [comment]
Have been trying this out and tried to implement it for the autodetect of version in graphite datasource. It sort of works, but the biggest issue as I see it is that the config_ctrl won't get any signaling that datasource have been updated - only when existing datasource are loaded or when a new datasources been created (but before onAfterSave hook) the constructor are called of config_ctrl.

In the graphite case I want that the version drop down should include the new autodetected version. Tried to implement so that config_ctrl could subscribe to events from ds_edit_ctrl, but didn't work out as good as I hoped.

Maybe my idea wasn't that good after all.

**Comment by svenklemm** [comment]
I could implement timescaledb autodetection with it and it worked, but I wasnt trying to modify any state in ds_edit_ctrl just in the datasource. I dont have a good idea for your use case atm but i'm gonna need that feature too later on. If the ds_edit_ctrl had an instance of the datasource as member then it should be possible because you could just reference that in your template, but I haven't verified that.

**Comment by torkelo** [comment]
It seems it's only called on update not on initial save. This seems very clunky (not your fault @svenklemm) due to how the proxy & ds edit controller works, and how Datasource instances needs a saved instance, and how datasourceSrv has a cache of all data sources for the org etc. 

Wonder if there was a way to do a bigger refactor that would allow test & auto detect stuff before save (even initial).

**Comment by marefr** [comment]
> Wonder if there was a way to do a bigger refactor that would allow test & auto detect stuff before save (even initial). 

That would be the most awesome and probably needed.

**Pull request closed without merging**